src/main/org/codehaus/groovy/control/ResolveVisitor.java

```diff
@@ -852,6 +858,21 @@ protected Expression transformPropertyExpression(PropertyExpression pe) {
         return ret;
     }
 
+    private boolean isVisibleNestedClass(ClassNode type, ClassNode ceType) {
+        if (!type.isRedirectNode()) return false;
+        ClassNode redirect = type.redirect();
+        // public or private
+        if (Modifier.isPublic(redirect.getModifiers()) || Modifier.isProtected(redirect.getModifiers())) return true;
+        // package local
+        PackageNode classPackage = ceType.getPackage();
+        PackageNode nestedPackage = redirect.getPackage();
+        if ((redirect.getModifiers() & (Modifier.PROTECTED | Modifier.PUBLIC | Modifier.PRIVATE)) == 0 &&
+                ((classPackage == null && nestedPackage == null) ||
+                        classPackage != null && nestedPackage != null && classPackage.equals(nestedPackage)))
+            return true;
+        return false;
```

Comment:
Comment says package local but it could be private at this point.  If you added an early exit `if (Modifier.isPrivate(redirect.getModifiers()) return false;` above the comment this would be package-private and the modifier check could be eliminated.

If that's the case then the return could be simplified:

```
return (classPackage == null && nestedPackage == null) || 
         (classPackage != null && classPackage.equals(nestedPackage))
```

I notice that `PackageNode` doesn't implement `equals` so it will end up being an `==` check.  I'm not that familiar with how the AST is built and if the same `PackageNode` is used for a given name so just wanted to bring that up.